src/Http/WebUtilities/src/FormPipeReader.cs

```diff
@@ -0,0 +1,342 @@
+            {
+                return GetDecodedString(ros.First.Span);
+            }
+
+            if (ros.Length < StackAllocThreshold)
+            {
+                Span<byte> buffer = stackalloc byte[(int)ros.Length];
+                ros.CopyTo(buffer);
+                return GetDecodedString(buffer);
+            }
+            else
+            {
+                var byteArray = ArrayPool<byte>.Shared.Rent((int)ros.Length);
+                Span<byte> buffer = byteArray.AsSpan(0, (int)ros.Length);
+
+                ros.CopyTo(buffer);
+                var decodedString = GetDecodedString(buffer);
+
+                ArrayPool<byte>.Shared.Return(byteArray);
+                return decodedString;
+            }
+        }
+
+        // Check that key/value constraints are met and appends value to accumulator.
+        private KeyValueAccumulator AppendAndVerify(ref KeyValueAccumulator accumulator, string decodedKey, string decodedValue)
+        {
+            if (decodedKey.Length > KeyLengthLimit)
+            {
+                throw new InvalidDataException($"Form key or value length limit {KeyLengthLimit} exceeded.");
+            }
+
+            if (decodedValue.Length > ValueLengthLimit)
+            {
+                throw new InvalidDataException($"Form key or value length limit {ValueLengthLimit} exceeded.");
+            }
+
+            accumulator.Append(decodedKey, decodedValue);
+
+            if (accumulator.ValueCount > ValueCountLimit)
+            {
+                throw new InvalidDataException($"Form value count limit {ValueCountLimit} exceeded.");
+            }
+
+            return accumulator;
+        }
+
+        private string GetDecodedString(in ReadOnlySpan<byte> readOnlySpan)
+        {
+            if (readOnlySpan.Length == 0)
+            {
+                return "";
+            }
+            else if (_encoding == Encoding.UTF8 || _encoding == Encoding.ASCII)
+            {
+                // UrlDecoder only works on UTF8 (and implicitly ASCII)
+
+                // We need to create a Span from a ReadOnlySpan. This cast is safe because the memory is still held by the pipe
+                // We will also create a string from it by the end of the function.
+                var span = MemoryMarshal.CreateSpan(ref Unsafe.AsRef(readOnlySpan[0]), readOnlySpan.Length);
+
+                var bytes = UrlDecoder.DecodeInPlace(span, isFormEncoding: true);
+                span = span.Slice(0, bytes);
+
+                return _encoding.GetString(span);
+            }
+            else
+            {
+                // Slow path for Unicode and other encodings.
+                // Just do raw string replacement.
+                var decodedString = _encoding.GetString(readOnlySpan);
+                decodedString = decodedString.Replace('+', ' ');
+                return Uri.UnescapeDataString(decodedString);
+            }
+        }
+
+        private ReadOnlySpan<byte> GetEqualsForEncoding()
```

Comment:
Just curious, why adding `GetEqualsForEncoding()` and `GetAndForEncoding()` and not initializing `_otherEqualEncoding` and `_otherAndEncoding` directly in the constructor? So it could be possible to remove the lines:
```
var equalsDelimiter = GetEqualsForEncoding();
var andDelimiter = GetAndForEncoding();
```